hummingbird.py

```diff
@@ -0,0 +1,979 @@
+          ff = "_".join(f.split("_")[:-1])
+        else:
+          ff = "_".join(f.split("_")[:-2])
+        values[f+"_max"] = (
+            np.max(measure_field[f], axis=0)[1] * self.scl_sampling_period)
+        values[f+"_min"] = (
+            np.min(measure_field[f], axis=0)[1] * self.scl_sampling_period)
+        values[f+"_worst"] = values[f+"_min"]
+        measure_idx = np.argmin(measure_field[f], axis=0)[1]
+        result[f+"_idx"] = measure_field[f][measure_idx][0]
+        svgwidth[f] = measure_field[f][measure_idx][1]
+        if values[f+"_worst"] >= spec_limit[ff]:
+          result[f] = 0
+        else:
+          result[f] = 1
+        result[f+"_margin"] = values[f+"_worst"] - spec_limit[ff]
+        result[f+"_percent"] = result[f+"_margin"] / spec_limit[ff] * 100
+
+    fields5 = [
+        "t_HD_DAT_rising_host", "t_HD_DAT_falling_host", "t_HD_DAT_rising_dev",
+        "t_HD_DAT_falling_dev"
+    ]
+    for f in fields5:
+      if self.requested_measurements[f+"_worst"] and measure_field[f]:
+        ff = "_".join(f.split("_")[:-2])
+        values[f+"_max"] = (
+            np.max(measure_field[f], axis=0)[1] * self.scl_sampling_period)
+        values[f+"_min"] = (
+            np.min(measure_field[f], axis=0)[1] * self.scl_sampling_period)
+        if spec_limit.get(ff) is None:
+          limit_max = np.inf
+        else:
+          limit_max = spec_limit[ff]
+        if values[f+"_min"] >= 0 and values[f+"_max"] <= limit_max:
+          result[f] = 0
+        else:
+          result[f] = 1
+        if values[f+"_min"] - 0 < limit_max - values[f+"_max"]:
+          values[f+"_worst"] = values[f+"_min"]
+          measure_idx = np.argmin(measure_field[f], axis=0)[1]
+          result[f+"_idx"] = measure_field[f][measure_idx][0]
+          svgwidth[f] = measure_field[f][measure_idx][1]
+          result[f+"_margin"] = values[f+"_min"] - 0
+        else:
+          values[f+"_worst"] = values[f+"_max"]
+          measure_idx = np.argmax(measure_field[f], axis=0)[1]
+          result[f+"_idx"] = measure_field[f][measure_idx][0]
+          svgwidth[f] = measure_field[f][measure_idx][1]
+          result[f+"_margin"] = limit_max - values[f+"_max"]
+        result[f+"_percent"] = result[f+"_margin"]/limit_max *100
+
+    fail = {}
+    if self.requested_measurements["spec"]:
+      fail = {param: result for (param, result) in result.items()
+              if ((result == 1) and ("_margin" not in param) and
+                  ("_percent" not in param) and ("_idx" not in param))}
+      passes = {param: result for (param, result) in result.items()
+                if ((result == 0) and ("_margin" not in param) and
+                    ("_percent" not in param) and ("_idx" not in param))}
+      num_pass = len(passes)
+      values["spec"] = len(fail)
+
+    uni_addr = list(set(addr_list))
+    uni_addr = [f"0x{int(addr, 2):02X}" for addr in uni_addr]
+    uni_addr = list(set(uni_addr))
```

Comment:
no need because duplicated addresses in `uni_addr` is already removed in line 907